src/tests/profiler/native/gcheapenumerationprofiler/gcheapenumerationprofiler.cpp

```diff
@@ -0,0 +1,237 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#include "gcheapenumerationprofiler.h"
+#include "../profilerstring.h"
+
+#if WIN32
+#define EXPORT
+#else // WIN32
+#define EXPORT __attribute__ ((visibility ("default")))
+#endif // WIN32
+
+GUID GCHeapEnumerationProfiler::GetClsid()
+{
+    // {8753F0E1-6D6D-4329-B8E1-334918869C15}
+	GUID clsid = { 0x8753f0e1, 0x6d6d, 0x4329,{ 0xb8, 0xe1, 0x33, 0x49, 0x18, 0x86, 0x9c, 0x15 } };
+	return clsid;
+}
+
+HRESULT GCHeapEnumerationProfiler::Initialize(IUnknown* pICorProfilerInfoUnk)
+{
+    Profiler::Initialize(pICorProfilerInfoUnk);
+    printf("GCHeapEnumerationProfiler::Initialize\n");
+
+    int eventMask = 0;
+    constexpr ULONG bufferSize = 1024;
+    ULONG envVarLen = 0;
+    WCHAR envVar[bufferSize];
+    HRESULT hr = S_OK;
+
+    if (FAILED(hr = pCorProfilerInfo->GetEnvironmentVariable(WCHAR("Set_Monitor_GC_Event_Mask"), bufferSize, &envVarLen, envVar)))
+    {
+        _failures++;
+        printf("FAIL: ICorProfilerInfo::GetEnvironmentVariable() failed hr=0x%x", hr);
+        return hr;
+    }
+    String envVarStr = envVar;
+    String expectedEnvVarStr = reinterpret_cast<const WCHAR*>(u"TRUE");
+    if (envVarStr == expectedEnvVarStr)
+    {
+        printf("Setting GarbageCollectionStarted event masks\n");
+        eventMask |= COR_PRF_MONITOR_GC;
+    }
+
+    if (FAILED(hr = pCorProfilerInfo->SetEventMask2(eventMask, COR_PRF_HIGH_MONITOR_NONE)))
+    {
+        _failures++;
+        printf("FAIL: ICorProfilerInfo::SetEventMask2() failed hr=0x%x", hr);
+    }
+
+    return hr;
+}
+
+HRESULT STDMETHODCALLTYPE GCHeapEnumerationProfiler::GarbageCollectionStarted(int cGenerations, BOOL generationCollected[], COR_PRF_GC_REASON reason)
+{
+    SHUTDOWNGUARD();
+    printf("GCHeapEnumerationProfiler::GarbageCollectionStarted\nSleeping for 10 seconds\n");
+    std::this_thread::sleep_for(std::chrono::seconds(10));
+    return S_OK;
```

Comment:
Lets validate that if you try to call EnumerateGCHeapObjects from inside CollectionStart/CollectionFinished that it doesn't work. I believe these callbacks are both no-trigger callbacks so I'd expect the calls to fail with CORPROF_E_UNSUPPORTED_CALL_SEQUENCE if you've added the additional kP2EETriggers check above.

Reply:
Calling EnumerateGCHeapObjects from inside GarbageCollectionStarted throws a Contract violation assert from the `CONTRACTL/CONTRACTL_END` scope before hitting the `PROFILER_TO_CLR_ENTRYPOINT_ASYNC_EX` macro and additional `!IsCalledAsynchronously() && !AreCallbackStateFlagsSet(COR_PRF_CALLBACKSTATE_IN_TRIGGERS_SCOPE)` check

Reply:
yeah I expect that test case would only work in a retail build. I thought we had some test attribute that did that but don't worry about including it if it isn't straigtforward.